core/module_specifier.rs

```diff
@@ -49,17 +109,50 @@ impl ModuleSpecifier {
   /// example is allowed here, whereas in import statements a leading "./" is
   /// required ("./foo.js"). This function is aware of the current working
   /// directory and returns an absolute URL.
-  pub fn resolve_root(
-    root_specifier: &str,
-  ) -> Result<ModuleSpecifier, url::ParseError> {
-    if let Ok(url) = Url::parse(root_specifier) {
-      Ok(ModuleSpecifier(url))
-    } else {
-      let cwd = std::env::current_dir().unwrap();
-      let base = Url::from_directory_path(cwd).unwrap();
-      let url = base.join(root_specifier)?;
-      Ok(ModuleSpecifier(url))
+  pub fn resolve_from_cwd(
+    specifier: &str,
+  ) -> Result<ModuleSpecifier, ResolveError> {
+    if let Ok(module_specifier) = ModuleSpecifier::resolve_absolute(specifier) {
+      return Ok(module_specifier);
+    }
+
+    // fallback to relative file path
+    // TODO: factor out as `normalize_path` method - it is used in /cli/deno_dir.rs and should
+    //  be applied to resolve https://github.com/denoland/deno/issues/1798
+    // HACK: `Url::from_directory_path` is used here because it normalizes the path.
+    // Joining `/dev/deno/" with "./tests" using `PathBuf` yields `/deno/dev/./tests/`.
+    // On the other hand joining `/dev/deno/" with "./tests" using `Url` yields "/dev/deno/tests"
+    // - and that's what we want.
+    // There exists similar method on `PathBuf` - `PathBuf.canonicalize`, but the problem
+    // is `canonicalize` resolves symlinks and we don't want that.
+    // We just want o normalize the path...
+    let specifier_path = PathBuf::from(specifier);
+    let cwd = std::env::current_dir().unwrap();
+    let path = cwd.join(specifier_path);
+    ensure_valid_prefix(&path)?;
+    let url =
+      Url::from_file_path(path).expect("PathBuf should be parseable as Url");
+    Ok(ModuleSpecifier(url))
+  }
+
+  /// Takes a string representing a path or URL to a module - must be absolute file path
+  /// or remote URL
+  pub fn resolve_absolute(
+    specifier: &str,
+  ) -> Result<ModuleSpecifier, ResolveError> {
+    use ResolveError::*;
+
+    // first check if specifier is an absolute path
+    let path = PathBuf::from(specifier);
```

Comment:
This doesn't check that the path is relative I think. 
At least, the existence of the `https://doc.rust-lang.org/std/path/struct.Path.html#method.is_relative` method suggests that it could be relative also.

Reply:
This method does not, but if it is relative path then on the next line `Url::from_file_path` will return error (`RelativeUrlWithoutBase`). I'll add comment to explain it better.

Reply:
Comment added